### PR TITLE
Make rules use iterators and whitespace assertions

### PR DESCRIPTION
**Pull request by TheSavior**
Should complete step one and two of #516.

Branched from #1014 to get the new differentLine assertion. That branch should be merged first (which will automatically remove a bunch of the commits from this PR)

### CODE REVIEW

**Comment by TheSavior** [comment]
Rebased from master

**Comment by mikesherov** [comment]
@TheSavior, will get to this shortly. Just want to make sure we address 1.11 bugs before landing this. Thanks again!

File: lib/rules/disallow-semicolons.js [hunk 1]
```diff
@@ -43,15 +43,15 @@ module.exports.prototype = {
...
-            .filter(function(token) {
-                return token.type === 'Punctuator' && token.value === ';';
-            })
-            .forEach(function(token) {
-                var nextToken = file.getNextToken(token);
-                if (!nextToken || nextToken.loc.end.line > token.loc.end.line) {
-                    errors.add('semicolons are disallowed at the end of a line.', token.loc.end);
-                }
-            });
+        file.iterateTokensByType('Punctuator', function(token) {
```
Comment (mikesherov) [inline lib/rules/disallow-semicolons.js]: there is now `iterateTokensByTypeAndValue`... please use it.

File: lib/token-assert.js [hunk 2]
```diff
@@ -96,11 +96,14 @@ TokenAssert.prototype.sameLine = function(options) {
...
 /**
- * Requires tokens to be on different lines.
+ * Requires tokens to be on different lines. If atLeast, atMost, and exactly, aren't specified, defaults to at least 1.
  *
  * @param {Object} options.token
  * @param {Object} options.nextToken
  * @param {Object} [options.message]
+ * @param {Object} [options.atLeast] At least how many lines the tokens are apart
+ * @param {Object} [options.atMost] At most how many lines the tokens are apart
+ * @param {object} [options.exactly] Exactly how many lines the tokens are apart
```
Comment (twolfson) [inline lib/token-assert.js]: Should we rename `exactly` to `lines` for consistency with `whitespaceBetween` (which uses `spaces`)?

Comment (TheSavior) [inline lib/token-assert.js]: Check out #1014. We have talked about the redesign of the API there.

Comment (twolfson) [inline lib/token-assert.js]: I saw https://github.com/jscs-dev/node-jscs/pull/1014#discussion_r24298315 which used `lines` and see that the current PR uses `exactly`. But I can't seem to find when the discussion about moving to `exactly``/atLeast`/`atMost`. Can you link me?

Comment (twolfson) [inline lib/token-assert.js]: For reference, I am coming from a PR about disallowing multiple spaces and added in a similar `atMost` option. We noticed the discrepency and figured we should comment.

#1059

**Comment by hzoo** [comment]
Since you're already doing

`lib/rules/require-padding-newlines-in-blocks.js`, you also could go ahead and do the `disallow` rule as well? Same with `lib/rules/require-spaces-inside-parenteheses.js` and the test!

**Comment by mdevils** [comment]
Amazing job!

**Comment by TheSavior** [comment]
@hzoo I thought I got all of them, so if there are any others that are still not using the iterators add them as comments here and I'll put them in this PR. Thanks!

**Comment by hzoo** [comment]
Ok checking and I got

- [x] lib/rules/disallow-padding-newlines-in-objects.js
- [ ] lib/rules/require-padding-newlines-in-objects.js
- [x] lib/rules/require-padding-newlines-in-blocks.js
- [ ] lib/rules/disallow-padding-newlines-in-blocks.js
- [x] lib/rules/require-spaces-inside-parenteheses.js
- [ ] lib/rules/disallow-spaces-inside-parenteheses.js

**Comment by mikesherov** [comment]
@TheSavior can you rebase this please? and note the I changed linesBetween to be 1 less than before.

**Comment by TheSavior** [comment]
Rebased but clearly rebased wrong due to tests failing on `differentLine`. I'll fix in a few hours, running out now.

File: lib/rules/disallow-padding-newlines-before-keywords.js [hunk 3]
```diff
@@ -95,12 +95,13 @@ module.exports.prototype = {
...
                 var prevToken = file.getPrevToken(token);
 
-                if (prevToken && token.loc.start.line - prevToken.loc.end.line > 1) {
-                    errors.add(
-                        'Keyword `' + token.value + '` should not have an empty line above it',
-                        token.loc.start.line,
-                        token.loc.start.column
-                    );
+                if (prevToken) {
+                    errors.assert.differentLine({
```
Comment (mikesherov) [inline lib/rules/disallow-padding-newlines-before-keywords.js]: This should now be `linesBetween`

File: lib/rules/require-padding-newlines-before-keywords.js [hunk 4]
```diff
@@ -107,13 +107,14 @@ module.exports.prototype = {
...
-                if (prevToken && prevToken.value !== '{' && prevToken.value !== ')' &&
-                    token.loc.start.line - prevToken.loc.end.line < 2) {
-                    errors.add(
-                        'Keyword `' + token.value + '` should have an empty line above it',
-                        token.loc.start.line,
-                        token.loc.start.column
-                    );
+                if (prevToken && prevToken.value !== '{' && prevToken.value !== ')') {
+
+                    errors.assert.differentLine({
```
Comment (mikesherov) [inline lib/rules/require-padding-newlines-before-keywords.js]: This should now be `linesBetween`

File: lib/rules/require-padding-newlines-in-blocks.js [hunk 5]
```diff
@@ -89,16 +89,22 @@ module.exports.prototype = {
             var openingBracket = file.getFirstNodeToken(node);
             var nextToken = file.getCommentAfterToken(openingBracket) || file.getNextToken(openingBracket);
 
-            if (nextToken.loc.start.line - openingBracket.loc.start.line < 2) {
-                errors.add('Expected a padding newline after opening curly brace', openingBracket.loc.end);
-            }
+            errors.assert.differentLine({
```
Comment (mikesherov) [inline lib/rules/require-padding-newlines-in-blocks.js]: This should now be `linesBetween`

File: lib/rules/require-padding-newlines-in-blocks.js [hunk 6]
```diff
@@ -89,16 +89,22 @@ module.exports.prototype = {
...
+                message: 'Expected a padding newline after opening curly brace'
+            });
 
             var closingBracket = file.getLastNodeToken(node);
             var prevToken = file.getCommentBeforeToken(closingBracket) || file.getPrevToken(closingBracket);
 
-            if (closingBracket.loc.start.line - prevToken.loc.start.line < 2) {
-                errors.add('Expected a padding newline before closing curly brace', prevToken.loc.end);
-            }
+            errors.assert.differentLine({
```
Comment (mikesherov) [inline lib/rules/require-padding-newlines-in-blocks.js]: This should now be `linesBetween`

File: test/token-assert.js [hunk 7]
```diff
@@ -359,6 +359,22 @@ describe('modules/token-assert', function() {
...
+            var onError = sinon.spy();
+            tokenAssert.on('error', onError);
+
+            var tokens = file.getTokens();
+            tokenAssert.differentLine({
+                token: tokens[0],
+                nextToken: tokens[1]
+            });
+
+            assert(!onError.calledOnce);
```
Comment (mikesherov) [inline test/token-assert.js]: This should now be `assert(!onError.called);`

**Comment by mikesherov** [comment]
Thanks again @TheSavior!

**Comment by TheSavior** [comment]
Thanks for finding a bunch of the broken pieces. Looks like the tests are passing.

**Comment by mikesherov** [comment]
Woooh!

**Comment by mikesherov** [comment]
@hzoo can you take the remaining ones that you identified from the list?

**Comment by hzoo** [comment]
Yeah

**Comment by mikesherov** [comment]
@hzoo Thanks!

**Comment by mikesherov** [comment]
@TheSavior thanks again!

**Pull request closed without merging**